File: native/ios/Sources/PatientExportBundle.swift
import Foundation

/// Validates the supported application collection profile, not arbitrary clinical FHIR.
enum PatientExportBundle {
    static func verifiedBytes(_ response: NativeResponse, identity: Identity,
                              contract: PatientExportContract, subjectField: String, maxBytes: Int) throws -> Data {
        do {
            guard identity.role == .patient, !identity.subject.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty,
                  identity.expiresAt > Date() else { throw MobileFailure.response }
            let bytes = try ExportIntegrity.verifiedBytes(response)
            guard maxBytes > 0, bytes.count <= maxBytes else { throw MobileFailure.response }
            let bundle = try object(bytes, depth: contract.maxJsonDepth)
            guard Set(bundle.keys) == ["resourceType", "type", "timestamp", "entry"],
                  bundle["resourceType"] as? String == "Bundle", bundle["type"] as? String == "collection",
                  let timestamp = bundle["timestamp"] as? String,
                  let entries = bundle["entry"] as? [[String: Any]], entries.count == contract.collections.count
            else { throw MobileFailure.response }
            let date = ISO8601DateFormatter()
            date.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
            if date.date(from: timestamp) == nil {
                date.formatOptions = [.withInternetDateTime]
                guard date.date(from: timestamp) != nil else { throw MobileFailure.response }
            }
            var identifiers = Set<String>(), collections = Set<String>()
            for entry in entries {
                guard Set(entry.keys) == ["fullUrl", "resource"], let resource = entry["resource"] as? [String: Any],
                      Set(resource.keys) == ["resourceType", "id", "contentType", "data"],
                      resource["resourceType"] as? String == "Binary", resource["contentType"] as? String == "application/json",
                      let id = resource["id"] as? String, UUID(uuidString: id)?.uuidString.lowercased() == id,
                      identifiers.insert(id).inserted, entry["fullUrl"] as? String == "urn:uuid:" + id,
                      let encoded = resource["data"] as? String, let decoded = Data(base64Encoded: encoded),
                      decoded.base64EncodedString() == encoded else { throw MobileFailure.response }
                let collection = try object(decoded, depth: contract.maxJsonDepth)
                guard Set(collection.keys) == ["collection", "records"], let name = collection["collection"] as? String,
                      contract.collections.contains(name), collections.insert(name).inserted else { throw MobileFailure.response }
                if name == contract.ownerCollection {
                    guard let patient = collection["records"] as? [String: Any],
                          patient[subjectField] as? String == identity.subject else { throw MobileFailure.response }
                } else if !(collection["records"] is [Any]) { throw MobileFailure.response }
            }
            guard collections == Set(contract.collections) else { throw MobileFailure.response }
            return bytes
        } catch { throw MobileFailure.response }
    }

    private static func object(_ data: Data, depth: Int) throws -> [String: Any] {
        var parser = StrictExportJSON(data: data)
        try parser.validate(maxDepth: depth)
        guard let object = try JSONSerialization.jsonObject(with: data) as? [String: Any] else { throw MobileFailure.response }
        return object
    }
}

/// RFC 8259 grammar checked before Foundation can discard duplicate object members.
private struct StrictExportJSON {
    private enum State { case objectStart, key, colon, value, objectEnd, arrayStart, arrayValue, arrayEnd }
    private struct Frame { var state: State; var keys = Set<String>() }
    private let bytes: [UInt8]
    private var index = 0
    private var stack: [Frame] = []
    init(data: Data) { bytes = Array(data) }

    mutating func validate(maxDepth: Int) throws {
        guard maxDepth > 0, String(bytes: bytes, encoding: .utf8) != nil else { throw MobileFailure.response }
        try value(maxDepth: maxDepth)
        while !stack.isEmpty {
            whitespace()
            let top = stack.count - 1
            switch stack[top].state {
            case .objectStart:
                if take(125) { stack.removeLast() } else { stack[top].state = .key }
            case .key:
                let key = try string()
                guard stack[top].keys.insert(key).inserted else { throw MobileFailure.response }
                stack[top].state = .colon
            case .colon:
                guard take(58) else { throw MobileFailure.response }; stack[top].state = .value
            case .value:
                stack[top].state = .objectEnd; try value(maxDepth: maxDepth)
            case .objectEnd:
                if take(125) { stack.removeLast() }
                else if take(44) { stack[top].state = .key }
                else { throw MobileFailure.response }
            case .arrayStart:
                if take(93) { stack.removeLast() } else { stack[top].state = .arrayValue }
            case .arrayValue:
                stack[top].state = .arrayEnd; try value(maxDepth: maxDepth)
            case .arrayEnd:
                if take(93) { stack.removeLast() }
                else if take(44) { stack[top].state = .arrayValue }
                else { throw MobileFailure.response }
            }
        }
        whitespace()
        guard index == bytes.count else { throw MobileFailure.response }
    }

    private mutating func value(maxDepth: Int) throws {
        whitespace()
        guard index < bytes.count else { throw MobileFailure.response }
        switch bytes[index] {
        case 123, 91:
            let state: State = bytes[index] == 123 ? .objectStart : .arrayStart
            guard stack.count < maxDepth else { throw MobileFailure.response }
            index += 1; stack.append(Frame(state: state))
        case 34: _ = try string()
        case 116: try literal("true")
        case 102: try literal("false")
        case 110: try literal("null")
        default:
            let start = index
            while index < bytes.count && ![9, 10, 13, 32, 44, 93, 125].contains(bytes[index]) { index += 1 }
            let number = String(decoding: bytes[start..<index], as: UTF8.self)
            guard number.range(of: "^-?(0|[1-9][0-9]*)(\\.[0-9]+)?([eE][+-]?[0-9]+)?$", options: .regularExpression) != nil,
                  let value = Double(number), value.isFinite else { throw MobileFailure.response }
        }
    }

    private mutating func string() throws -> String {
        let start = index
        guard take(34) else { throw MobileFailure.response }
        while index < bytes.count {
            let byte = bytes[index]; index += 1
            if byte == 34 { return try JSONDecoder().decode(String.self, from: Data(bytes[start..<index])) }
            guard byte >= 32 else { throw MobileFailure.response }
            if byte == 92 {
                guard index < bytes.count else { throw MobileFailure.response }
                let escape = bytes[index]; index += 1
                if escape == 117 {
                    let unit = try hexUnit()
                    if (0xD800...0xDBFF).contains(unit) {
                        guard take(92), take(117), (0xDC00...0xDFFF).contains(try hexUnit()) else { throw MobileFailure.response }
                    } else if (0xDC00...0xDFFF).contains(unit) { throw MobileFailure.response }
                } else if ![34, 92, 47, 98, 102, 110, 114, 116].contains(escape) { throw MobileFailure.response }
            }
        }
        throw MobileFailure.response
    }

    private mutating func hexUnit() throws -> UInt16 {
        guard index + 4 <= bytes.count else { throw MobileFailure.response }
        let digits = bytes[index..<index + 4]
        guard digits.allSatisfy({ (48...57).contains($0) || (65...70).contains($0) || (97...102).contains($0) }),
              let value = UInt16(String(decoding: digits, as: UTF8.self), radix: 16) else { throw MobileFailure.response }
        index += 4; return value
    }
    private mutating func literal(_ text: String) throws {
        for byte in text.utf8 { guard take(byte) else { throw MobileFailure.response } }
    }
    private mutating func take(_ byte: UInt8) -> Bool {
        guard index < bytes.count, bytes[index] == byte else { return false }
        index += 1; return true
    }
    private mutating func whitespace() {
        while index < bytes.count && [9, 10, 13, 32].contains(bytes[index]) { index += 1 }
    }
}

File: native/android/app/src/main/java/com/mediconnect/nativeapp/PatientExportBundle.kt
package com.mediconnect.nativeapp

import com.google.gson.JsonArray
import com.google.gson.JsonElement
import com.google.gson.JsonNull
import com.google.gson.JsonObject
import com.google.gson.JsonPrimitive
import com.google.gson.Strictness
import com.google.gson.stream.JsonReader
import com.google.gson.stream.JsonToken
import java.io.StringReader
import java.nio.ByteBuffer
import java.nio.charset.CodingErrorAction
import java.nio.charset.StandardCharsets
import java.time.Instant
import java.util.Base64
import java.util.UUID

/** The supported portability profile wraps application collections in FHIR Binary resources. */
object PatientExportBundle {
    fun verifiedBytes(response: NativeResponse, identity: Identity, contract: PatientExportContract,
                      subjectField: String, maxBytes: Long): ByteArray {
        try {
            require(identity.role == Role.PATIENT && identity.subject.isNotBlank() && identity.expiresAt > Instant.now())
            val bytes = ExportIntegrity.verifiedBytes(response)
            require(maxBytes > 0 && bytes.size <= maxBytes)
            val bundle = StrictExportJson.read(bytes, contract.maxJsonDepth).asJsonObject
            require(bundle.keySet() == setOf("resourceType", "type", "timestamp", "entry"))
            require(bundle.string("resourceType") == "Bundle" && bundle.string("type") == "collection")
            Instant.parse(bundle.string("timestamp"))
            val entries = bundle.getAsJsonArray("entry")
            require(entries.size() == contract.collections.size)
            val identifiers = mutableSetOf<String>()
            val collections = mutableSetOf<String>()
            for (element in entries) {
                val entry = element.asJsonObject
                require(entry.keySet() == setOf("fullUrl", "resource"))
                val resource = entry.getAsJsonObject("resource")
                require(resource.keySet() == setOf("resourceType", "id", "contentType", "data"))
                require(resource.string("resourceType") == "Binary" && resource.string("contentType") == "application/json")
                val id = resource.string("id")
                require(UUID.fromString(id).toString() == id && identifiers.add(id))
                require(entry.string("fullUrl") == "urn:uuid:$id")
                val encoded = resource.string("data")
                val decoded = Base64.getDecoder().decode(encoded)
                require(Base64.getEncoder().encodeToString(decoded) == encoded)
                val collection = StrictExportJson.read(decoded, contract.maxJsonDepth).asJsonObject
                require(collection.keySet() == setOf("collection", "records"))
                val name = collection.string("collection")
                require(name in contract.collections && collections.add(name))
                val records = collection.get("records")
                if (name == contract.ownerCollection) {
                    require(records.asJsonObject.string(subjectField) == identity.subject)
                } else require(records.isJsonArray)
            }
            require(collections == contract.collections)
            return bytes
        } catch (_: Exception) { throw ApiFailure() }
    }

    private fun JsonObject.string(name: String): String {
        val value = get(name)
        require(value != null && value.isJsonPrimitive && value.asJsonPrimitive.isString)
        return value.asString
    }
}

/** Strict UTF-8/JSON with duplicate-member and nesting checks, without recursive parsing. */
private object StrictExportJson {
    private class Frame(val value: JsonElement, var name: String? = null)

    fun read(bytes: ByteArray, maxDepth: Int): JsonElement {
        require(maxDepth > 0)
        val text = StandardCharsets.UTF_8.newDecoder().onMalformedInput(CodingErrorAction.REPORT)
            .onUnmappableCharacter(CodingErrorAction.REPORT).decode(ByteBuffer.wrap(bytes)).toString()
        val stack = mutableListOf<Frame>()
        var root: JsonElement? = null
        fun attach(value: JsonElement) {
            val parent = stack.lastOrNull()
            if (parent == null) { require(root == null); root = value }
            else if (parent.value.isJsonArray) parent.value.asJsonArray.add(value)
            else {
                val name = requireNotNull(parent.name)
                require(!parent.value.asJsonObject.has(name))
                parent.value.asJsonObject.add(name, value); parent.name = null
            }
        }
        JsonReader(StringReader(text)).use { reader ->
            reader.strictness = Strictness.STRICT
            while (true) {
                when (reader.peek()) {
                    JsonToken.BEGIN_OBJECT -> {
                        reader.beginObject(); val value = JsonObject(); attach(value); stack.add(Frame(value))
                        require(stack.size <= maxDepth)
                    }
                    JsonToken.BEGIN_ARRAY -> {
                        reader.beginArray(); val value = JsonArray(); attach(value); stack.add(Frame(value))
                        require(stack.size <= maxDepth)
                    }
                    JsonToken.END_OBJECT -> { reader.endObject(); require(stack.removeAt(stack.lastIndex).name == null) }
                    JsonToken.END_ARRAY -> { reader.endArray(); stack.removeAt(stack.lastIndex) }
                    JsonToken.NAME -> {
                        val name = scalarString(reader.nextName()); val parent = stack.last()
                        require(parent.name == null && !parent.value.asJsonObject.has(name)); parent.name = name
                    }
                    JsonToken.STRING -> attach(JsonPrimitive(scalarString(reader.nextString())))
                    JsonToken.NUMBER -> { val value = reader.nextDouble(); require(value.isFinite()); attach(JsonPrimitive(value)) }
                    JsonToken.BOOLEAN -> attach(JsonPrimitive(reader.nextBoolean()))
                    JsonToken.NULL -> { reader.nextNull(); attach(JsonNull.INSTANCE) }
                    JsonToken.END_DOCUMENT -> break
                    else -> throw ApiFailure()
                }
            }
        }
        require(stack.isEmpty())
        return requireNotNull(root)
    }

    private fun scalarString(value: String): String {
        var index = 0
        while (index < value.length) {
            val character = value[index++]
            if (Character.isHighSurrogate(character)) {
                require(index < value.length && Character.isLowSurrogate(value[index++]))
            } else require(!Character.isLowSurrogate(character))
        }
        return value
    }
}
